fhir-persistence-blob-app/README.md

```diff
@@ -0,0 +1,225 @@
+
+The following examples use PostgreSQL as the database type, but the tool also supports db2 and derby as options.
+
+To identify orphan records without deleting anything, run:
+```
+java -jar fhir-persistence-blob-app-*-cli.jar \
+    --fhir-config-dir /path/to/wlp/usr/servers/defaultServer \
+    --tenant-id default \
+    --ds-id default \
+    --db-properties database.properties \
+    --db-type postgresql \
+    --reconcile \
+    --dry-run \
+    --max-scan-seconds 600 
+```
+
+To identify and delete orphan records, run:
+```
+java -jar fhir-persistence-blob-app-*-cli.jar \
+    --fhir-config-dir /path/to/wlp/usr/servers/defaultServer \
+    --tenant-id default \
+    --ds-id default \
+    --db-properties database.properties \
+    --db-type postgresql \
+    --reconcile \
+    --confirm \
+    --max-scan-seconds 600 
+```
+
+To continue an earlier scan and delete which didn't complete, use the last continuation token value reported in the log of the previous run then specify when running the command again:
+
+```
+java -jar fhir-persistence-blob-app-*-cli.jar \
+    --fhir-config-dir /path/to/wlp/usr/servers/defaultServer \
+    --tenant-id default \
+    --ds-id default \
+    --db-properties database.properties \
+    --db-type postgresql  \
+    --reconcile \
+    --confirm \
+    --max-scan-seconds 600 \
+    --continuation-token "<token>"
+```
+The continuation token can be identified in the log by searching for the string `__CONTINUATION_TOKEN__ =`. The application logs an INFO message after each page of records is processed, so use the last occurrence found in the log. The value of the continuation token is opaque and only meaningful to the Azure Blob service.
+
+## Development
+
+For development, payload offload can be configured to point to a local `azurite` container which emulates the API of the Azure Blob store service. Run the following command to start a local `azurite` container:
+
+```
+podman run -d -p 10000:10000 \
+    -v ./data/azurite:/data:z \
+    mcr.microsoft.com/azure-storage/azurite \
+    azurite-blob --blobHost 0.0.0.0 --blobPort 10000
+```
+
+More info on using `azurite` for development can be found in the official documents here: https://docs.microsoft.com/en-us/azure/storage/common/storage-use-azurite?tabs=docker-hub.
+
+### Checking Stored Values
+
+When new resources are ingested, the class `com.ibm.fhir.persistence.blob.BlobStorePayload` logs a `FINE` message which can be seen in the Liberty Profile `trace.log` file:
+
+```
+Payload storage path: 97/my-blob-5/7/b0a2fb75-1ef8-45c8-b5d7-658350fb45cc
+```
+Where:
+| Value | Meaning |
+| ----- | ------- |
+|        97 | `resource_type_id` primary key from RDBMS `resource_types` table |
+| my-blob-5 | resource logical-id |
```

Comment:
just double-checking:  all the allowed chars for FHIR Resource.id are valid in an azure blob path, right?
```
[A-Za-z0-9\-\.]{1,64}
```

Reply:
Almost. There is one issue: https://docs.microsoft.com/en-us/rest/api/storageservices/naming-and-referencing-containers--blobs--and-metadata#blob-names

> Avoid blob names that end with a dot (.), a forward slash (/), or a sequence or combination of the two. No path segments should end with a dot (.).

FHIR resource names can end with a dot (although in practice this is unlikely out of convention). Some sort of escape encoding will be required to support this. The escape character can be one that isn't a valid FHIR id character.

Reply:
Added encode/decode for the offload blob path so any `.` is replaced with a `#` (the latter not being a valid character for FHIR logical identifiers, making it a simple replace without any complicated escaping logic.

Reply:
actually, maybe its worth mentioning that in the README here...just in case anyone needs to work with the resources in azure blob directly?

Reply:
done